src/DotLiquid/Util/NumericConverter.cs

```diff
@@ -0,0 +1,190 @@
+        /// Based on the promotion table at
+        /// https://docs.microsoft.com/en-us/dotnet/standard/base-types/conversion-tables
+        /// </summary>
+        public static readonly IReadOnlyDictionary<Type, Type[]> NumericTypePromotions = new ReadOnlyDictionary<Type, Type[]>(new Dictionary<Type, Type[]>
+        {
+            { typeof(Byte), new Type[] { typeof(UInt16), typeof(Int16), typeof(UInt32), typeof(Int32), typeof(UInt64), typeof(Int64), typeof(Decimal), typeof(Single), typeof(Double) } },
+            { typeof(SByte), new Type[] { typeof(Int16), typeof(Int32), typeof(Int64), typeof(Decimal), typeof(Single), typeof(Double) } },
+            { typeof(Int16), new Type[] { typeof(Int32), typeof(Int64), typeof(Decimal), typeof(Single), typeof(Double) } },
+            { typeof(UInt16), new Type[] { typeof(UInt32), typeof(Int32), typeof(UInt64), typeof(Int64), typeof(Decimal), typeof(Single), typeof(Double) } },
+            { typeof(Char), new Type[] { typeof(UInt16), typeof(UInt32), typeof(Int32), typeof(UInt64), typeof(Int64), typeof(Decimal), typeof(Single), typeof(Double) } },
+            { typeof(Int32), new Type[] { typeof(Int64), typeof(Decimal), typeof(Single), typeof(Double) } },
+            { typeof(UInt32), new Type[] { typeof(Int64), typeof(UInt64), typeof(Decimal), typeof(Single), typeof(Double) } },
+            { typeof(Int64), new Type[] { typeof(Decimal), typeof(Single), typeof(Double) } },
+            { typeof(UInt64), new Type[] { typeof(Decimal), typeof(Single), typeof(Double) } },
+            { typeof(Single), new Type[] { typeof(Double) } },
+            { typeof(Decimal), new Type[] { typeof(Single), typeof(Double) } },
+            { typeof(Double), new Type[] { } },
+        });
+
+        /// <summary>
+        /// Get the return type for an operation between the two numbers of the specified types
+        /// </summary>
+        /// <param name="left">Type of the left parameter</param>
+        /// <param name="right">Type of the left parameter</param>
+        /// <returns>The return type</returns>
+        /// <exception cref="ArgumentException">Thrown if a parameter is not a supported numeric type.</exception>
+        public static Type GetBinaryResultType(Type left, Type right)
+        {
+            if (left == right)
+                return left;
+
+            if (!NumericConverter.NumericTypePromotions.TryGetValue(left, out Type[] leftTypes))
+                throw new System.ArgumentException("Argument is not numeric", nameof(left));
+            if (!NumericConverter.NumericTypePromotions.TryGetValue(right, out Type[] rightTypes))
+                throw new System.ArgumentException("Argument is not numeric", nameof(right));
+
+            // Test left to right promotion
+            if (rightTypes.Contains(left))
+                return left;
+            if (leftTypes.Contains(right))
+                return right;
+            return rightTypes.First(t => leftTypes.Contains(t));

```

Comment:
```suggestion
            return rightTypes.First(_type => leftTypes.Contains(_type));

```
Using English terms to disambiguate and for readability, with _ prefix to avoid clash with language reserved keyword.